### PR TITLE
Function input decoder

### PR DESCRIPTION
**Pull request by banteg**
### What was wrong?

There was no trivial method to decode function input parameters.

### How was it fixed?

This adds `Contract.decode_function_input` which is a counterpart to `Contract.encodeABI`.

```python
>>> data = '338b5dea00000000000000000000000042d6622dece394b54999fbd73d108123806f6a180000000000000000000000000000000000000000000000000000000000000032'
>>> contract.decode_function_input(data)
(<Function depositToken(address,uint256)>,
 {'token': '0x42d6622dece394b54999fbd73d108123806f6a18', 'amount': 50})
```

Works like a charm ✨ 

#### Cute Animal Picture

My bunny loves Ethereum

### CODE REVIEW

File: tests/core/contracts/test_contract_method_abi_decoding.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        (
+            ABI_ADDRESS,
+            '0x4767be6c000000000000000000000000ffffffffffffffffffffffffffffffffffffffff',
+            'addressfunc',
+            [],
+            {'addressarg': '0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF'},
+        ),
+    ),
+)
+def test_contract_abi_decoding(web3, abi, data, method, arguments, expected):
```
Comment (carver) [inline tests/core/contracts/test_contract_method_abi_decoding.py]: Do you want to add a test against `arguments`? It's currently unused.

File: docs/contracts.rst [hunk 2]
```diff
@@ -762,3 +762,17 @@ For example:
...
+.. py:classmethod:: Contract.decode_function_input(data)
+
+    Decodes function input data and returns :py:class:`ContractFunction` and decoded parameters as :py:class:`dict`.
+
+    .. code-block:: python
+
+        >>> data = '338b5dea00000000000000000000000042d6622dece394b54999fbd73d108123806f6a180000000000000000000000000000000000000000000000000000000000000032'
+        >>> contract.decode_function_input(data)
+        (<Function depositToken(address,uint256)>,
+         {'token': '0x42d6622dece394b54999fbd73d108123806f6a18', 'amount': 50})
```
[the remark this answers is not in the document]
Comment (banteg) [reply docs/contracts.rst]: ENS example actually sounds great, I'll update the docs.

**Comment by carver** [comment]
Unrelated test failure: occasionally parity integration tests are flaky.

**Comment by carver** [comment]
Very nice, thanks for the contribution!

**Pull request merged**